integration/container/daemon_linux_test.go

```diff
@@ -0,0 +1,96 @@
+package container
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/integration-cli/daemon"
+	"golang.org/x/sys/unix"
+)
+
+// This is a regression test for #36145
+// It ensures that a container can be started when the daemon was improperly
+// shutdown when the daemon is brought back up.
+//
+// The regression is due to improper error handling preventing a container from
+// being restored and as such have the resources cleaned up.
+//
+// To test this, we need to kill dockerd, then kill both the containerd-shim and
+// the container process, then start dockerd back up and attempt to start the
+// container again.
+func TestContainerStartOnDaemonRestart(t *testing.T) {
+	t.Parallel()
+
+	d := daemon.New(t, "", "dockerd", daemon.Config{})
+	d.StartWithBusybox(t, "--iptables=false")
+	defer d.Stop(t)
+
+	client, err := d.NewClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx := context.Background()
+	c, err := client.ContainerCreate(ctx,
+		&container.Config{
+			Image: "busybox",
+			Cmd:   []string{"top"},
+		},
+		nil,
+		nil,
+		"",
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.ContainerRemove(ctx, c.ID, types.ContainerRemoveOptions{Force: true})
+
+	if err := client.ContainerStart(ctx, c.ID, types.ContainerStartOptions{}); err != nil {
+		t.Fatal(err)
+	}
+
+	inspect, err := client.ContainerInspect(ctx, c.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if inspect.State.Pid <= 1 {
+		t.Fatalf("got unexpected container PID: %d", inspect.State.Pid)
+	}
+
+	// Get the pid of the containerd-shim
+	statB, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/stat", inspect.State.Pid))
```

Comment:
I think this (line 62-79) should be extracted into a helper so it doesn't distract from the test logic:

```go
ppid := getContainerdShimPID(t, inspect.State.Pid)
```